src/java.base/share/classes/java/lang/StackTraceElement.java

```diff
@@ -406,16 +406,14 @@ public String toString() {
     public boolean equals(Object obj) {
         if (obj==this)
             return true;
-        if (!(obj instanceof StackTraceElement))
-            return false;
-        StackTraceElement e = (StackTraceElement)obj;
-        return Objects.equals(classLoaderName, e.classLoaderName) &&
-            Objects.equals(moduleName, e.moduleName) &&
-            Objects.equals(moduleVersion, e.moduleVersion) &&
-            e.declaringClass.equals(declaringClass) &&
-            e.lineNumber == lineNumber &&
-            Objects.equals(methodName, e.methodName) &&
-            Objects.equals(fileName, e.fileName);
+        return (obj instanceof StackTraceElement e)
+                && Objects.equals(classLoaderName, e.classLoaderName)
+                && Objects.equals(moduleName, e.moduleName)
+                && Objects.equals(moduleVersion, e.moduleVersion)
+                && e.declaringClass.equals(declaringClass)
```

Comment:
testing the declaring class and the line number should be done first given they are primitive values, it will be a little more efficient if two StackTraceElement are not equals and one is using non interned String.
```
  return (obj instanceof StackTraceElement e)
     && e.lineNumber == lineNumber
     && e.declaringClass == declaringClass
     && ...
```

Reply:
Hi Rémi,
There seems to be a deeper issue here - Patrick pointed that out to me - the specification of equals above speaks of comparing class names where the actual implementation compares classes. Maybe the specification should be updated - but that would be better done in a separate issue with CSR etc... Maybe we can do the optimization you suggest at the same time?

Reply:
`declaringClass` is a string representing the class name (not the `Class` object).   The variable name indeed causes confusion.

Reply:
Doh. My mistake. Was thinking about `StackFrame`. Thanks Mandy!

Reply:
I've reordered the checks as suggested and the updated code can be found in commit b89027c